heapster-saw/doc/tutorial/tutorial.md

```diff
@@ -0,0 +1,623 @@
+```
+
+Notice that this is a memory-safety specification, not a correctness
+predicate. We can reason about the correctness of the function once we
+extract it as a functional program in Coq.
+
+In the next sections, we shall define the predicates necessary to
+write this specification in Heapster.
+
+##### Defining permission predicates
+
+One of the simplest Heapster commands is `heapster_define_perm`, which defines
+a new named permission which can then be used in Heapster types. As an
+example, the following defines the permission which describes a 64-bit integer
+value:
+
+```
+heapster_define_perm env "int64" " " "llvmptr 64" "exists x:bv 64.eq(llvmword(x))"
+```
+
+The first argument is the Heapster environment, the second is its
+name, the third is its arguments (of which there are none), the fourth
+is the type of value that the permission applies to, and the fifth is
+its definition. the new permission is created and added to the
+environment.
+
+Unfortunately, there is currently no way to print the newly defined
+permissions. If you try to print the environment (`print env`) at this
+point, you will only see the `llvm` definitions. We might add
+functionality for showing permissions in the future.
+
+The permission predicate for lists is a little more complicated
+because it requires a recursive definition. To define
+[permissions](doc/Permissions.md) which can describe unbounded data
+structures, you can use the `heapster_define_recursive_perm`
+command. As an example, here is how to describe a linked list of
+64-bit words using this command:
+
+```
+heapster_define_recursive_perm
+  env
+  "List64"
+  "rw:rwmodality"
+  "llvmptr 64"
+  ["eq(llvmword(0))", "ptr((rw,0) |-> int64<>) * ptr((rw,8) |-> List64<rw>)"]
+  "List (Vec 64 Bool)"
+  "foldList (Vec 64 Bool)"
+  "unfoldList (Vec 64 Bool)";
+```
+
+Its first four arguments are the same as for `heapster_define_perm`,
+its fifth argument contains its different inductive cases (in this
+case, a `List64` is either a null pointer, or a pointer to an `Int64`
+and another `List64`), and its final three arguments are its
+translation into SAW core. Here the SAW core definitions used are from
+the SAW core prelude which are included by default when Heapster
+autogenerates a a module for you. If you need new SAW core
+definitions, you will need to use the following command instead of
+`heapster_init_env`:
+
+```
+env <- heapster_init_env_from_file "my_file.sawcore" "my_file.bc";
+```
+
+See this [additional documentation](doc/Permissions.md) for a
+reference on the syntax and meaning of heapster permissions.
+
+#### 5. Writing a SAW script to type-check your code with respect to the sepcification
+   
+We begin by converting the pen-and-paper specification of `is_elem`
+into a Heapster specification. First, every heapster type begins with
+the declaration of a ghost environment. Our specification doesn't use
+any ghost variable so that is `().` Moreover, for every predicate we
```

Comment:
This explanation of every predicate requiring a read-only or read-write modality is not quite right. The reason we write `List64<R>` here is because `List64` was defined to take a single argument of type `rwmodality`. Read-write access to a list would require the type `List64<W>`. You can't use `<>` in place of `<R>` because the former is 0 arguments and the latter is 1.